// playcanvas@2.19.7/build/playcanvas.dbg/src/framework/handlers/html.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
import { http } from "../../platform/net/http.js";
import { ResourceHandler } from "./handler.js";
class HtmlHandler extends ResourceHandler {
  constructor(app) {
    super(app, "html");
    /**
     * TextDecoder for decoding binary data.
     *
     * @type {TextDecoder|null}
     * @private
     */
    __publicField(this, "decoder", null);
  }
  load(url, callback) {
    if (typeof url === "string") {
      url = {
        load: url,
        original: url
      };
    }
    http.get(url.load, {
      retry: this.maxRetries > 0,
      maxRetries: this.maxRetries
    }, (err, response) => {
      if (!err) {
        callback(null, response);
      } else {
        callback(`Error loading html resource: ${url.original} [${err}]`);
      }
    });
  }
  /**
   * Parses raw DataView and returns string.
   *
   * @param {DataView} data - The raw data as a DataView
   * @returns {string} The parsed resource data.
   */
  openBinary(data) {
    this.decoder ?? (this.decoder = new TextDecoder("utf-8"));
    return this.decoder.decode(data);
  }
}
export {
  HtmlHandler
};
