// playcanvas@2.19.7/build/playcanvas.dbg/src/scene/geometry/sphere-geometry.js
import { calculateTangents } from "./geometry-utils.js";
import { Geometry } from "./geometry.js";
class SphereGeometry extends Geometry {
  /**
   * Create a new SphereGeometry instance.
   *
   * By default, the constructor creates a sphere centered on the object space origin with a radius
   * of 0.5 and 16 segments in both longitude and latitude. The sphere is created with UVs in the
   * range of 0 to 1.
   *
   * @param {object} [opts] - Options object.
   * @param {number} [opts.radius] - The radius of the sphere. Defaults to 0.5.
   * @param {number} [opts.latitudeBands] - The number of divisions along the latitudinal axis of
   * the sphere. Defaults to 16.
   * @param {number} [opts.longitudeBands] - The number of divisions along the longitudinal axis of
   * the sphere. Defaults to 16.
   * @param {boolean} [opts.calculateTangents] - Generate tangent information. Defaults to false.
   * @example
   * const geometry = new pc.SphereGeometry({
   *     radius: 1,
   *     latitudeBands: 32,
   *     longitudeBands: 32
   * });
   */
  constructor(opts = {}) {
    super();
    const radius = opts.radius ?? 0.5;
    const latitudeBands = opts.latitudeBands ?? 16;
    const longitudeBands = opts.longitudeBands ?? 16;
    const positions = [];
    const normals = [];
    const uvs = [];
    const indices = [];
    for (let lat = 0; lat <= latitudeBands; lat++) {
      const theta = lat * Math.PI / latitudeBands;
      const sinTheta = Math.sin(theta);
      const cosTheta = Math.cos(theta);
      for (let lon = 0; lon <= longitudeBands; lon++) {
        const phi = lon * 2 * Math.PI / longitudeBands - Math.PI / 2;
        const sinPhi = Math.sin(phi);
        const cosPhi = Math.cos(phi);
        const x = cosPhi * sinTheta;
        const y = cosTheta;
        const z = sinPhi * sinTheta;
        const u = 1 - lon / longitudeBands;
        const v = 1 - lat / latitudeBands;
        positions.push(x * radius, y * radius, z * radius);
        normals.push(x, y, z);
        uvs.push(u, 1 - v);
      }
    }
    for (let lat = 0; lat < latitudeBands; ++lat) {
      for (let lon = 0; lon < longitudeBands; ++lon) {
        const first = lat * (longitudeBands + 1) + lon;
        const second = first + longitudeBands + 1;
        indices.push(first + 1, second, first);
        indices.push(first + 1, second + 1, second);
      }
    }
    this.positions = positions;
    this.normals = normals;
    this.uvs = uvs;
    this.uvs1 = uvs;
    this.indices = indices;
    if (opts.calculateTangents) {
      this.tangents = calculateTangents(positions, normals, uvs, indices);
    }
  }
}
export {
  SphereGeometry
};
