// playcanvas@2.19.7/build/playcanvas.dbg/src/framework/graphics/render-pass-picker.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
import { DebugGraphics } from "../../platform/graphics/debug-graphics.js";
import { BlendState } from "../../platform/graphics/blend-state.js";
import { RenderPass } from "../../platform/graphics/render-pass.js";
import { SHADER_PICK, SHADER_DEPTH_PICK } from "../../scene/constants.js";
const tempMeshInstances = [];
const lights = [[], [], []];
const defaultShadowAtlasParams = new Float32Array(2);
class RenderPassPicker extends RenderPass {
  constructor(device, renderer) {
    super(device);
    /** @type {BindGroup[]} */
    __publicField(this, "viewBindGroups", []);
    /** @type {BlendState} */
    __publicField(this, "blendState", BlendState.NOBLEND);
    /** @type {CameraComponent} */
    __publicField(this, "camera");
    /** @type {Scene} */
    __publicField(this, "scene");
    /** @type {Layer[]} */
    __publicField(this, "layers");
    /** @type {Map<number, MeshInstance | GSplatComponent>} */
    __publicField(this, "mapping");
    /** @type {boolean} */
    __publicField(this, "depth");
    /** @type {number[]} */
    __publicField(this, "_qualifiedLayerIndices", []);
    /** @type {Map<number, MeshInstance|null>} */
    __publicField(this, "_pickMeshInstances", /* @__PURE__ */ new Map());
    this.renderer = renderer;
  }
  destroy() {
    this.viewBindGroups.forEach((bg) => {
      bg.defaultUniformBuffer.destroy();
      bg.destroy();
    });
    this.viewBindGroups.length = 0;
  }
  /**
   * @param {CameraComponent} camera - The camera component used for picking.
   * @param {Scene} scene - The scene to pick from.
   * @param {Layer[]} layers - The layers to pick from.
   * @param {Map<number, MeshInstance | GSplatComponent>} mapping - Map to store ID to object mappings.
   * @param {boolean} depth - Whether to render depth information.
   */
  update(camera, scene, layers, mapping, depth) {
    this.camera = camera;
    this.scene = scene;
    this.layers = layers;
    this.mapping = mapping;
    this.depth = depth;
    if (scene.clusteredLightingEnabled) {
      this.emptyWorldClusters = this.renderer.worldClustersAllocator.empty;
    }
  }
  // Filter qualifying layers and prepare gsplat pick mesh instances for the compute-based
  // renderer. The execute() loop iterates the pre-built list instead of re-filtering.
  before() {
    this._qualifiedLayerIndices.length = 0;
    this._pickMeshInstances.clear();
    const { camera, scene, layers } = this;
    const srcLayers = scene.layers.layerList;
    const subLayerEnabled = scene.layers.subLayerEnabled;
    const gsplatDirector = this.renderer.gsplatDirector;
    const pickerWidth = this.renderTarget?.width ?? 1;
    const pickerHeight = this.renderTarget?.height ?? 1;
    for (let i = 0; i < srcLayers.length; i++) {
      const srcLayer = srcLayers[i];
      if (layers && layers.indexOf(srcLayer) < 0) continue;
      if (!srcLayer.enabled || !subLayerEnabled[i]) continue;
      if (!srcLayer.camerasSet.has(camera.camera)) continue;
      this._qualifiedLayerIndices.push(i);
      if (gsplatDirector) {
        const pickMI = gsplatDirector.prepareForPicking(camera.camera, pickerWidth, pickerHeight, srcLayer);
        if (pickMI) {
          this._pickMeshInstances.set(i, pickMI);
        }
      }
    }
  }
  execute() {
    const device = this.device;
    const { renderer, camera, scene, mapping, renderTarget } = this;
    const srcLayers = scene.layers.layerList;
    const isTransparent = scene.layers.subLayerList;
    for (const i of this._qualifiedLayerIndices) {
      const srcLayer = srcLayers[i];
      const transparent = isTransparent[i];
      DebugGraphics.pushGpuMarker(device, `${srcLayer.name}(${transparent ? "TRANSP" : "OPAQUE"})`);
      if (srcLayer._clearDepthBuffer) {
        renderer.clear(camera.camera, false, true, false);
      }
      const meshInstances = srcLayer.meshInstances;
      for (let j = 0; j < meshInstances.length; j++) {
        const meshInstance = meshInstances[j];
        if (meshInstance.pick && meshInstance.transparent === transparent) {
          tempMeshInstances.push(meshInstance);
          mapping.set(meshInstance.id, meshInstance);
        }
      }
      const pickMI = this._pickMeshInstances.get(i);
      if (pickMI) {
        tempMeshInstances.push(pickMI);
      }
      if (scene.gsplat.enableIds) {
        const placements = srcLayer.gsplatPlacements;
        for (let j = 0; j < placements.length; j++) {
          const placement = placements[j];
          const component = placement.node?.gsplat;
          if (component) {
            mapping.set(placement.id, component);
          }
        }
      }
      if (tempMeshInstances.length > 0) {
        const clusteredLightingEnabled = scene.clusteredLightingEnabled;
        if (clusteredLightingEnabled) {
          const lightClusters = this.emptyWorldClusters;
          lightClusters.activate();
        }
        renderer.setCameraUniforms(camera.camera, renderTarget);
        renderer.dispatchGlobalLights(scene);
        device.scope.resolve("shadowAtlasParams").setValue(defaultShadowAtlasParams);
        if (device.supportsUniformBuffers) {
          renderer.initViewBindGroupFormat(clusteredLightingEnabled);
          renderer.setupViewUniformBuffers(this.viewBindGroups, renderer.viewUniformFormat, renderer.viewBindGroupFormat, null);
        }
        const shaderPass = this.depth ? SHADER_DEPTH_PICK : SHADER_PICK;
        renderer.renderForward(camera.camera, renderTarget, tempMeshInstances, lights, shaderPass, (meshInstance) => {
          device.setBlendState(this.blendState);
        });
        tempMeshInstances.length = 0;
      }
      DebugGraphics.popGpuMarker(device);
    }
  }
}
export {
  RenderPassPicker
};
