// playcanvas@2.19.7/build/playcanvas.dbg/src/platform/graphics/webgpu/webgpu-query-set.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
import { DebugHelper } from "../../../core/debug.js";
class WebgpuQuerySet {
  constructor(device, isTimestamp, capacity) {
    /**
     * @type {GPUQuerySet}
     */
    __publicField(this, "querySet");
    __publicField(this, "stagingBuffers", []);
    __publicField(this, "activeStagingBuffer", null);
    /** @type {number} */
    __publicField(this, "bytesPerSlot");
    this.device = device;
    this.capacity = capacity;
    this.bytesPerSlot = isTimestamp ? 8 : 4;
    const wgpu = device.wgpu;
    this.querySet = wgpu.createQuerySet({
      type: isTimestamp ? "timestamp" : "occlusion",
      count: capacity
    });
    DebugHelper.setLabel(this.querySet, `QuerySet-${isTimestamp ? "Timestamp" : "Occlusion"}`);
    this.queryBuffer = wgpu.createBuffer({
      size: this.bytesPerSlot * capacity,
      usage: GPUBufferUsage.QUERY_RESOLVE | GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST
    });
    DebugHelper.setLabel(this.queryBuffer, "QueryGpuBuffer");
  }
  destroy() {
    this.device.deferDestroy(this.querySet);
    this.querySet = null;
    this.device.deferDestroy(this.queryBuffer);
    this.queryBuffer = null;
    this.activeStagingBuffer = null;
    this.stagingBuffers.forEach((stagingBuffer) => {
      stagingBuffer.destroy();
    });
    this.stagingBuffers = null;
  }
  getStagingBuffer() {
    let stagingBuffer = this.stagingBuffers.pop();
    if (!stagingBuffer) {
      stagingBuffer = this.device.wgpu.createBuffer({
        size: this.queryBuffer.size,
        usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ
      });
      DebugHelper.setLabel(this.queryBuffer, "QueryStagingBuffer");
    }
    return stagingBuffer;
  }
  resolve(count) {
    const device = this.device;
    const commandEncoder = device.getCommandEncoder();
    commandEncoder.resolveQuerySet(this.querySet, 0, count, this.queryBuffer, 0);
    const activeStagingBuffer = this.getStagingBuffer();
    this.activeStagingBuffer = activeStagingBuffer;
    commandEncoder.copyBufferToBuffer(this.queryBuffer, 0, activeStagingBuffer, 0, this.bytesPerSlot * count);
  }
  request(count, renderVersion) {
    const stagingBuffer = this.activeStagingBuffer;
    this.activeStagingBuffer = null;
    return stagingBuffer.mapAsync(GPUMapMode.READ).then(() => {
      const srcTimings = new BigInt64Array(stagingBuffer.getMappedRange());
      const timings = [];
      for (let i = 0; i < count; i++) {
        timings.push(Number(srcTimings[i * 2 + 1] - srcTimings[i * 2]) * 1e-6);
      }
      stagingBuffer.unmap();
      this.stagingBuffers?.push(stagingBuffer);
      return {
        renderVersion,
        timings
      };
    });
  }
}
export {
  WebgpuQuerySet
};
