// playcanvas@2.19.7/build/playcanvas/src/scene/graphics/radix-sort/compute-radix-sort-multipass.js
import { Vec2 } from "../../../core/math/vec2.js";
import { Compute } from "../../../platform/graphics/compute.js";
import { Shader } from "../../../platform/graphics/shader.js";
import { StorageBuffer } from "../../../platform/graphics/storage-buffer.js";
import { BindGroupFormat, BindStorageBufferFormat, BindUniformBufferFormat } from "../../../platform/graphics/bind-group-format.js";
import { UniformBufferFormat, UniformFormat } from "../../../platform/graphics/uniform-buffer-format.js";
import { BUFFERUSAGE_COPY_SRC, BUFFERUSAGE_COPY_DST, SHADERLANGUAGE_WGSL, SHADERSTAGE_COMPUTE, UNIFORMTYPE_UINT } from "../../../platform/graphics/constants.js";
import { PrefixSumKernel } from "../prefix-sum-kernel.js";
import { radixSort4bitSource } from "../../shader-lib/wgsl/chunks/radix-sort/compute-radix-sort-4bit.js";
import { radixSortReorderSource } from "../../shader-lib/wgsl/chunks/radix-sort/compute-radix-sort-reorder.js";
import { ComputeRadixSortBase } from "./compute-radix-sort-base.js";
const WORKGROUP_SIZE_X = 16;
const WORKGROUP_SIZE_Y = 16;
const THREADS_PER_WORKGROUP = WORKGROUP_SIZE_X * WORKGROUP_SIZE_Y;
const ELEMENTS_PER_THREAD = 8;
const ELEMENTS_PER_WORKGROUP = THREADS_PER_WORKGROUP * ELEMENTS_PER_THREAD;
const BITS_PER_PASS = 4;
const BUCKET_COUNT = 16;
class ComputeRadixSortMultipass extends ComputeRadixSortBase {
	_workgroupCount = 0;
	_allocatedWorkgroupCount = 0;
	_blockSums = null;
	_prefixSumKernel = null;
	_dispatchSize = new Vec2(1, 1);
	_histogramBindGroupFormat = null;
	_reorderBindGroupFormat = null;
	_uniformBufferFormat = null;
	_passes = [];
	constructor(device, indirect = false) {
		super(device, indirect);
		const info = this._indirectInfo;
		info[0] = 1;
		info[1] = ELEMENTS_PER_WORKGROUP;
		this._uniformBufferFormat = new UniformBufferFormat(device, [
			new UniformFormat("workgroupCount", UNIFORMTYPE_UINT),
			new UniformFormat("elementCount", UNIFORMTYPE_UINT)
		]);
		const histogramEntries = [
			new BindStorageBufferFormat("input", SHADERSTAGE_COMPUTE, true),
			new BindStorageBufferFormat("block_sums", SHADERSTAGE_COMPUTE, false),
			new BindUniformBufferFormat("uniforms", SHADERSTAGE_COMPUTE)
		];
		const reorderEntries = [
			new BindStorageBufferFormat("inputKeys", SHADERSTAGE_COMPUTE, true),
			new BindStorageBufferFormat("outputKeys", SHADERSTAGE_COMPUTE, false),
			new BindStorageBufferFormat("prefix_block_sum", SHADERSTAGE_COMPUTE, true),
			new BindStorageBufferFormat("inputValues", SHADERSTAGE_COMPUTE, true),
			new BindStorageBufferFormat("outputValues", SHADERSTAGE_COMPUTE, false),
			new BindUniformBufferFormat("uniforms", SHADERSTAGE_COMPUTE)
		];
		if (indirect) {
			histogramEntries.push(new BindStorageBufferFormat("sortElementCount", SHADERSTAGE_COMPUTE, true));
			reorderEntries.push(new BindStorageBufferFormat("sortElementCount", SHADERSTAGE_COMPUTE, true));
		}
		this._histogramBindGroupFormat = new BindGroupFormat(device, histogramEntries);
		this._reorderBindGroupFormat = new BindGroupFormat(device, reorderEntries);
	}
	destroy() {
		this._destroyBuffers();
		this._destroyPasses();
		this._histogramBindGroupFormat?.destroy();
		this._reorderBindGroupFormat?.destroy();
		this._histogramBindGroupFormat = null;
		this._reorderBindGroupFormat = null;
		this._uniformBufferFormat = null;
		super.destroy();
	}
	_destroyPasses() {
		for (const pass of this._passes) {
			pass.histogramCompute.shader?.destroy();
			pass.reorderCompute.shader?.destroy();
		}
		this._passes.length = 0;
		this._numBits = 0;
	}
	_destroyBuffers() {
		this._destroyPingPongBuffers();
		this._blockSums?.destroy();
		this._prefixSumKernel?.destroy();
		this._blockSums = null;
		this._prefixSumKernel = null;
		this._workgroupCount = 0;
		this._allocatedWorkgroupCount = 0;
	}
	get radixBits() {
		return BITS_PER_PASS;
	}
	_createPasses(numBits, hasInitialValues, skipLastPassKeyWrite) {
		this._destroyPasses();
		this._numBits = numBits;
		this._hasInitialValues = hasInitialValues;
		this._skipLastPassKeyWrite = skipLastPassKeyWrite;
		const numPasses = numBits / BITS_PER_PASS;
		const suffix = this._indirect ? "-Indirect" : "";
		for (let pass = 0; pass < numPasses; pass++) {
			const bitOffset = pass * BITS_PER_PASS;
			const isFirstPass = pass === 0 && !hasInitialValues;
			const isLastPass = skipLastPassKeyWrite && pass === numPasses - 1;
			const histogramShader = this._createShader(
				`RadixSort4bit-Histogram${suffix}-${bitOffset}`,
				radixSort4bitSource,
				bitOffset,
				false,
				false,
				this._histogramBindGroupFormat
			);
			const reorderShader = this._createShader(
				`RadixSort4bit-Reorder${suffix}-${bitOffset}`,
				radixSortReorderSource,
				bitOffset,
				isFirstPass,
				isLastPass,
				this._reorderBindGroupFormat
			);
			const histogramCompute = new Compute(this.device, histogramShader, `RadixSort4bit-Histogram${suffix}-${bitOffset}`);
			const reorderCompute = new Compute(this.device, reorderShader, `RadixSort4bit-Reorder${suffix}-${bitOffset}`);
			this._passes.push({ histogramCompute, reorderCompute });
		}
	}
	_allocateBuffers(elementCount, numBits, hasInitialValues, skipLastPassKeyWrite, forceRealloc = false) {
		const effectiveCount = Math.max(elementCount, this.capacity);
		const allocWorkgroupCount = Math.ceil(effectiveCount / ELEMENTS_PER_WORKGROUP);
		const currentWorkgroupCount = Math.max(1, Math.ceil(elementCount / ELEMENTS_PER_WORKGROUP));
		const buffersNeedRealloc = forceRealloc || allocWorkgroupCount !== this._allocatedWorkgroupCount || !this._keys0;
		const passesNeedRecreate = numBits !== this._numBits || hasInitialValues !== this._hasInitialValues || skipLastPassKeyWrite !== this._skipLastPassKeyWrite;
		if (buffersNeedRealloc) {
			this._destroyBuffers();
			this._allocatedWorkgroupCount = allocWorkgroupCount;
			this.capacity = effectiveCount;
			const blockSumSize = BUCKET_COUNT * allocWorkgroupCount * 4;
			this._allocatePingPongElementBuffers(effectiveCount);
			this._blockSums = new StorageBuffer(this.device, blockSumSize, BUFFERUSAGE_COPY_SRC | BUFFERUSAGE_COPY_DST);
			this._prefixSumKernel = new PrefixSumKernel(this.device);
		}
		this._workgroupCount = currentWorkgroupCount;
		Compute.calcDispatchSize(currentWorkgroupCount, this._dispatchSize, this.device.limits.maxComputeWorkgroupsPerDimension || 65535);
		this._prefixSumKernel.resize(this._blockSums, BUCKET_COUNT * currentWorkgroupCount);
		if (passesNeedRecreate) {
			this._createPasses(numBits, hasInitialValues, skipLastPassKeyWrite);
		}
	}
	_createShader(name, source, currentBit, isFirstPass, isLastPass, bindGroupFormat) {
		const cdefines = /* @__PURE__ */ new Map();
		cdefines.set("{WORKGROUP_SIZE_X}", WORKGROUP_SIZE_X);
		cdefines.set("{WORKGROUP_SIZE_Y}", WORKGROUP_SIZE_Y);
		cdefines.set("{THREADS_PER_WORKGROUP}", THREADS_PER_WORKGROUP);
		cdefines.set("{ELEMENTS_PER_THREAD}", ELEMENTS_PER_THREAD);
		cdefines.set("{CURRENT_BIT}", currentBit);
		cdefines.set("{IS_FIRST_PASS}", isFirstPass ? 1 : 0);
		cdefines.set("{IS_LAST_PASS}", isLastPass ? 1 : 0);
		if (this._indirect) {
			cdefines.set("USE_INDIRECT_SORT", "");
		}
		return new Shader(this.device, {
			name,
			shaderLanguage: SHADERLANGUAGE_WGSL,
			cshader: source,
			cdefines,
			computeBindGroupFormat: bindGroupFormat,
			computeUniformBufferFormats: { uniforms: this._uniformBufferFormat }
		});
	}
	sort(keysBuffer, elementCount, numBits = 16, initialValues, skipLastPassKeyWrite = false, destructiveKeys = false) {
		return this._execute(keysBuffer, elementCount, numBits, -1, null, initialValues, skipLastPassKeyWrite, destructiveKeys);
	}
	sortIndirect(keysBuffer, maxElementCount, numBits, sortSlotBase, sortElementCountBuffer, initialValues, skipLastPassKeyWrite = false, destructiveKeys = false) {
		return this._execute(keysBuffer, maxElementCount, numBits, sortSlotBase, sortElementCountBuffer, initialValues, skipLastPassKeyWrite, destructiveKeys);
	}
	_execute(keysBuffer, elementCount, numBits, sortSlotBase, sortElementCountBuffer, initialValues, skipLastPassKeyWrite = false, destructiveKeys = false) {
		this._elementCount = elementCount;
		const hasInitialValues = !!initialValues;
		const prevDestructiveKeys = this._destructiveKeys;
		this._destructiveKeys = destructiveKeys;
		this._allocateBuffers(elementCount, numBits, hasInitialValues, skipLastPassKeyWrite, destructiveKeys !== prevDestructiveKeys);
		if (destructiveKeys) {
			this._keys1 = keysBuffer;
		}
		const device = this.device;
		const numPasses = numBits / BITS_PER_PASS;
		const suffix = this._indirect ? "-Indirect" : "";
		let currentKeys = keysBuffer;
		let currentValues = initialValues ?? this._values0;
		let nextKeys = this._keys0;
		let nextValues = this._values1;
		for (let pass = 0; pass < numPasses; pass++) {
			const { histogramCompute, reorderCompute } = this._passes[pass];
			const isLastPass = pass === numPasses - 1;
			if (this._indirect) {
				this._blockSums.clear();
			}
			histogramCompute.setParameter("input", currentKeys);
			histogramCompute.setParameter("block_sums", this._blockSums);
			histogramCompute.setParameter("workgroupCount", this._workgroupCount);
			histogramCompute.setParameter("elementCount", elementCount);
			if (this._indirect) {
				histogramCompute.setParameter("sortElementCount", sortElementCountBuffer);
				histogramCompute.setupIndirectDispatch(sortSlotBase);
			} else {
				histogramCompute.setupDispatch(this._dispatchSize.x, this._dispatchSize.y, 1);
			}
			device.computeDispatch([histogramCompute], `RadixSort4bit-Histogram${suffix}`);
			this._prefixSumKernel.dispatch(device);
			reorderCompute.setParameter("inputKeys", currentKeys);
			reorderCompute.setParameter("outputKeys", nextKeys);
			reorderCompute.setParameter("prefix_block_sum", this._blockSums);
			reorderCompute.setParameter("inputValues", currentValues);
			reorderCompute.setParameter("outputValues", nextValues);
			reorderCompute.setParameter("workgroupCount", this._workgroupCount);
			reorderCompute.setParameter("elementCount", elementCount);
			if (this._indirect) {
				reorderCompute.setParameter("sortElementCount", sortElementCountBuffer);
				reorderCompute.setupIndirectDispatch(sortSlotBase);
			} else {
				reorderCompute.setupDispatch(this._dispatchSize.x, this._dispatchSize.y, 1);
			}
			device.computeDispatch([reorderCompute], `RadixSort4bit-Reorder${suffix}`);
			if (!isLastPass) {
				currentKeys = nextKeys;
				nextKeys = currentKeys === this._keys0 ? this._keys1 : this._keys0;
				const tempValues = currentValues;
				currentValues = nextValues;
				nextValues = tempValues;
			}
		}
		return this.sortedIndices;
	}
}
export {
	ComputeRadixSortMultipass
};
