// playcanvas@2.19.7/build/playcanvas.dbg/src/platform/graphics/transform-feedback.js
import { Debug } from "../../core/debug.js";
import { BUFFER_GPUDYNAMIC, PRIMITIVE_POINTS } from "./constants.js";
import { VertexBuffer } from "./vertex-buffer.js";
import { DebugGraphics } from "./debug-graphics.js";
import { Shader } from "./shader.js";
import { ShaderDefinitionUtils } from "./shader-definition-utils.js";
class TransformFeedback {
  /**
   * Create a new TransformFeedback instance.
   *
   * @param {VertexBuffer} inputBuffer - The input vertex buffer.
   * @param {VertexBuffer} [outputBuffer] - The optional output buffer.
   * If not specified, a buffer with parameters matching the input buffer will be created.
   * @param {number} [usage] - The optional usage type of the output vertex buffer. Can be:
   *
   * - {@link BUFFER_STATIC}
   * - {@link BUFFER_DYNAMIC}
   * - {@link BUFFER_STREAM}
   * - {@link BUFFER_GPUDYNAMIC}
   *
   * Defaults to {@link BUFFER_GPUDYNAMIC} (which is recommended for continuous update).
   */
  constructor(inputBuffer, outputBuffer, usage = BUFFER_GPUDYNAMIC) {
    if (outputBuffer !== void 0 && !(outputBuffer instanceof VertexBuffer)) {
      Debug.deprecated("Such a constructor that takes the second parameter usage is deprecated.");
      usage = outputBuffer;
      outputBuffer = void 0;
    }
    this.device = inputBuffer.device;
    const gl = this.device.gl;
    const outVB = outputBuffer ?? inputBuffer;
    Debug.assert(
      outVB.format.interleaved || outVB.format.elements.length <= 1,
      outputBuffer ? "Output vertex buffer used by TransformFeedback needs to be interleaved." : "Input vertex buffer used by TransformFeedback needs to be interleaved."
    );
    if (usage === BUFFER_GPUDYNAMIC && outVB.usage !== usage) {
      gl.bindBuffer(gl.ARRAY_BUFFER, outVB.impl.bufferId);
      gl.bufferData(gl.ARRAY_BUFFER, outVB.storage, gl.DYNAMIC_COPY);
    }
    this._inputBuffer = inputBuffer;
    this._destroyOutputBuffer = !outputBuffer;
    this._outputBuffer = outputBuffer ?? new VertexBuffer(inputBuffer.device, inputBuffer.format, inputBuffer.numVertices, {
      usage,
      data: inputBuffer.storage
    });
  }
  /**
   * Creates a transform feedback ready vertex shader from code.
   *
   * @param {GraphicsDevice} graphicsDevice - The graphics device used by the renderer.
   * @param {string} vertexCode - Vertex shader code. Should contain output variables starting with "out_" or feedbackVaryings.
   * @param {string} name - Unique name for caching the shader.
   * @param {string[]} [feedbackVaryings] - A list of shader output variable names that will be captured.
   * @returns {Shader} A shader to use in the process() function.
   */
  static createShader(graphicsDevice, vertexCode, name, feedbackVaryings) {
    return new Shader(graphicsDevice, ShaderDefinitionUtils.createDefinition(graphicsDevice, {
      name,
      vertexCode,
      feedbackVaryings,
      useTransformFeedback: true,
      fragmentCode: "void main(void) {gl_FragColor = vec4(0.0);}"
    }));
  }
  /**
   * Destroys the transform feedback helper object.
   */
  destroy() {
    if (this._destroyOutputBuffer) {
      this._outputBuffer.destroy();
    }
  }
  /**
   * Runs the specified shader on the input buffer, writes results into the new buffer, then
   * optionally swaps input/output.
   *
   * @param {Shader} shader - A vertex shader to run. Should be created with
   * {@link TransformFeedback.createShader}.
   * @param {boolean} [swap] - Swap input/output buffer data. Useful for continuous buffer
   * processing. Default is true.
   */
  process(shader, swap = true) {
    const device = this.device;
    DebugGraphics.pushGpuMarker(device, "TransformFeedback");
    const oldRt = device.getRenderTarget();
    device.setRenderTarget(null);
    device.updateBegin();
    device.setVertexBuffer(this._inputBuffer);
    device.setRaster(false);
    device.setTransformFeedbackBuffer(this._outputBuffer);
    device.setShader(shader);
    device.draw({
      type: PRIMITIVE_POINTS,
      base: 0,
      baseVertex: 0,
      count: this._inputBuffer.numVertices,
      indexed: false
    });
    device.setTransformFeedbackBuffer(null);
    device.setRaster(true);
    device.updateEnd();
    device.setRenderTarget(oldRt);
    DebugGraphics.popGpuMarker(device);
    if (swap) {
      Debug.call(() => {
        if (this._inputBuffer.format !== this._outputBuffer.format) {
          Debug.warnOnce("Trying to swap buffers with different formats.");
        }
      });
      let tmp = this._inputBuffer.impl.bufferId;
      this._inputBuffer.impl.bufferId = this._outputBuffer.impl.bufferId;
      this._outputBuffer.impl.bufferId = tmp;
      tmp = this._inputBuffer.impl.vao;
      this._inputBuffer.impl.vao = this._outputBuffer.impl.vao;
      this._outputBuffer.impl.vao = tmp;
    }
  }
  /**
   * The current input buffer.
   *
   * @type {VertexBuffer}
   */
  get inputBuffer() {
    return this._inputBuffer;
  }
  /**
   * The current output buffer.
   *
   * @type {VertexBuffer}
   */
  get outputBuffer() {
    return this._outputBuffer;
  }
}
export {
  TransformFeedback
};
