// playcanvas@2.19.7/build/playcanvas.dbg/src/scene/graphics/render-pass-radix-sort-reorder.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
import { SEMANTIC_POSITION, SHADERLANGUAGE_GLSL, SHADERLANGUAGE_WGSL } from "../../platform/graphics/constants.js";
import { RenderPassShaderQuad } from "./render-pass-shader-quad.js";
import { ShaderUtils } from "../shader-lib/shader-utils.js";
import { ShaderChunks } from "../shader-lib/shader-chunks.js";
import glslRadixSortReorderPS from "../shader-lib/glsl/chunks/radix-sort/radix-sort-reorder.js";
import wgslRadixSortReorderPS from "../shader-lib/wgsl/chunks/radix-sort/radix-sort-reorder.js";
class RenderPassRadixSortReorder extends RenderPassShaderQuad {
  /**
   * @param {GraphicsDevice} device - The graphics device.
   * @param {boolean} sourceLinear - Whether to read from linear-layout source texture.
   * @param {boolean} outputLinear - Whether to output indices in linear layout.
   * @param {number} bitsPerStep - Bits per radix step (usually 4).
   * @param {number} groupSize - Log2 of group size (usually 4 for 16 elements).
   * @param {number} currentBit - Current bit offset for this pass.
   */
  constructor(device, sourceLinear, outputLinear, bitsPerStep, groupSize, currentBit) {
    super(device);
    /**
     * Whether this pass reads from linear-layout source texture (first pass).
     */
    __publicField(this, "sourceLinear", false);
    /**
     * Whether to output indices in linear layout.
     */
    __publicField(this, "outputLinear", false);
    /**
     * Bits per radix step (usually 4).
     */
    __publicField(this, "bitsPerStep", 0);
    /**
     * Log2 of group size (usually 4 for 16 elements).
     */
    __publicField(this, "groupSize", 0);
    /**
     * Current bit offset for this pass.
     */
    __publicField(this, "currentBit", 0);
    /**
     * Dynamic params updated per frame.
     *
     * @type {{elementCount: number, imageElementsLog2: number, imageSize: number}}
     * @private
     */
    __publicField(this, "_dynamicParams", { elementCount: 0, imageElementsLog2: 0, imageSize: 0 });
    this.sourceLinear = sourceLinear;
    this.outputLinear = outputLinear;
    this.bitsPerStep = bitsPerStep;
    this.groupSize = groupSize;
    this.currentBit = currentBit;
    ShaderChunks.get(device, SHADERLANGUAGE_GLSL).set("radixSortReorderPS", glslRadixSortReorderPS);
    ShaderChunks.get(device, SHADERLANGUAGE_WGSL).set("radixSortReorderPS", wgslRadixSortReorderPS);
    const defines = /* @__PURE__ */ new Map();
    if (sourceLinear) {
      defines.set("SOURCE_LINEAR", "");
    }
    if (outputLinear) {
      defines.set("OUTPUT_LINEAR", "");
    }
    let shaderName = "RadixSortReorderShader";
    if (sourceLinear) shaderName += "SourceLinear";
    if (outputLinear) shaderName += "OutputLinear";
    this.shader = ShaderUtils.createShader(device, {
      uniqueName: shaderName,
      attributes: { aPosition: SEMANTIC_POSITION },
      vertexChunk: "quadVS",
      fragmentChunk: "radixSortReorderPS",
      fragmentDefines: defines,
      fragmentOutputTypes: ["uvec4", "uvec4"]
      // MRT: keys (uint) and indices (uint)
    });
    this.keysTextureId = device.scope.resolve("keysTexture");
    if (!sourceLinear) {
      this.indicesTextureId = device.scope.resolve("indicesTexture");
    }
    this.prefixSumsId = device.scope.resolve("prefixSums");
    this.bitsPerStepId = device.scope.resolve("bitsPerStep");
    this.groupSizeId = device.scope.resolve("groupSize");
    this.elementCountId = device.scope.resolve("elementCount");
    this.imageElementsLog2Id = device.scope.resolve("imageElementsLog2");
    this.currentBitId = device.scope.resolve("currentBit");
    this.imageSizeId = device.scope.resolve("imageSize");
  }
  /**
   * Sets the keys texture to read from.
   *
   * @param {Texture} keysTexture - The keys texture (R32U).
   */
  setKeysTexture(keysTexture) {
    this._keysTexture = keysTexture;
  }
  /**
   * Sets the indices texture to read from.
   *
   * @param {Texture} indicesTexture - The indices texture (R32U).
   */
  setIndicesTexture(indicesTexture) {
    this._indicesTexture = indicesTexture;
  }
  /**
   * Sets the prefix sums texture.
   *
   * @param {Texture} prefixSums - The prefix sums texture (R32F with mipmaps).
   */
  setPrefixSumsTexture(prefixSums) {
    this._prefixSums = prefixSums;
  }
  /**
   * Sets dynamic parameters (called each frame).
   *
   * @param {number} elementCount - Number of elements to sort.
   * @param {number} imageElementsLog2 - Log2 of total texture elements.
   * @param {number} imageSize - Size of the internal texture (power of 2).
   */
  setDynamicParams(elementCount, imageElementsLog2, imageSize) {
    this._dynamicParams.elementCount = elementCount;
    this._dynamicParams.imageElementsLog2 = imageElementsLog2;
    this._dynamicParams.imageSize = imageSize;
  }
  execute() {
    this.keysTextureId.setValue(this._keysTexture);
    if (!this.sourceLinear) {
      this.indicesTextureId.setValue(this._indicesTexture);
    }
    this.prefixSumsId.setValue(this._prefixSums);
    this.bitsPerStepId.setValue(this.bitsPerStep);
    this.groupSizeId.setValue(this.groupSize);
    this.elementCountId.setValue(this._dynamicParams.elementCount);
    this.imageElementsLog2Id.setValue(this._dynamicParams.imageElementsLog2);
    this.currentBitId.setValue(this.currentBit);
    this.imageSizeId.setValue(this._dynamicParams.imageSize);
    super.execute();
  }
}
export {
  RenderPassRadixSortReorder
};
