// playcanvas@2.19.7/build/playcanvas.prf/src/platform/graphics/webgpu/webgpu-texture.js
import { TRACEID_RENDER_QUEUE } from "../../../core/constants.js";
import { math } from "../../../core/math/math.js";
import {
	pixelFormatInfo,
	isCompressedPixelFormat,
	getPixelFormatArrayType,
	ADDRESS_REPEAT,
	ADDRESS_CLAMP_TO_EDGE,
	ADDRESS_MIRRORED_REPEAT,
	PIXELFORMAT_RGBA16F,
	PIXELFORMAT_RGBA32F,
	PIXELFORMAT_DEPTHSTENCIL,
	SAMPLETYPE_UNFILTERABLE_FLOAT,
	SAMPLETYPE_DEPTH,
	FILTER_NEAREST,
	FILTER_LINEAR,
	FILTER_NEAREST_MIPMAP_NEAREST,
	FILTER_NEAREST_MIPMAP_LINEAR,
	FILTER_LINEAR_MIPMAP_NEAREST,
	FILTER_LINEAR_MIPMAP_LINEAR,
	isIntegerPixelFormat,
	SAMPLETYPE_INT,
	SAMPLETYPE_UINT,
	BUFFERUSAGE_READ,
	BUFFERUSAGE_COPY_DST
} from "../constants.js";
import { TextureUtils } from "../texture-utils.js";
import { gpuTextureFormats } from "./constants.js";
const gpuAddressModes = [];
gpuAddressModes[ADDRESS_REPEAT] = "repeat";
gpuAddressModes[ADDRESS_CLAMP_TO_EDGE] = "clamp-to-edge";
gpuAddressModes[ADDRESS_MIRRORED_REPEAT] = "mirror-repeat";
const gpuFilterModes = [];
gpuFilterModes[FILTER_NEAREST] = { level: "nearest", mip: "nearest" };
gpuFilterModes[FILTER_LINEAR] = { level: "linear", mip: "nearest" };
gpuFilterModes[FILTER_NEAREST_MIPMAP_NEAREST] = { level: "nearest", mip: "nearest" };
gpuFilterModes[FILTER_NEAREST_MIPMAP_LINEAR] = { level: "nearest", mip: "linear" };
gpuFilterModes[FILTER_LINEAR_MIPMAP_NEAREST] = { level: "linear", mip: "nearest" };
gpuFilterModes[FILTER_LINEAR_MIPMAP_LINEAR] = { level: "linear", mip: "linear" };
const dummyUse = (thingOne) => {
};
class WebgpuTexture {
	gpuTexture;
	view;
	samplers = [];
	desc;
	format;
	viewCache = /* @__PURE__ */ new Map();
	constructor(texture) {
		this.texture = texture;
		this.format = gpuTextureFormats[texture.format];
		this.create(texture.device);
	}
	create(device) {
		const texture = this.texture;
		const wgpu = device.wgpu;
		const numLevels = texture.numLevels;
		if (isCompressedPixelFormat(texture.format) && (texture.width % 4 !== 0 || texture.height % 4 !== 0)) {
			texture._width = math.roundUp(texture.width, 4);
			texture._height = math.roundUp(texture.height, 4);
		}
		this.desc = {
			size: {
				width: texture.width,
				height: texture.height,
				depthOrArrayLayers: texture.cubemap ? 6 : texture.array ? texture.arrayLength : 1
			},
			format: this.format,
			mipLevelCount: numLevels,
			sampleCount: 1,
			dimension: texture.volume ? "3d" : "2d",
			// TODO: use only required usage flags
			// COPY_SRC - probably only needed on render target textures, to support copyRenderTarget (grab pass needs it)
			// RENDER_ATTACHMENT - needed for mipmap generation
			usage: GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.COPY_SRC | (isCompressedPixelFormat(texture.format) ? 0 : GPUTextureUsage.RENDER_ATTACHMENT) | (texture.storage ? GPUTextureUsage.STORAGE_BINDING : 0)
		};
		this.gpuTexture = wgpu.createTexture(this.desc);
		let viewDescr;
		if (this.texture.format === PIXELFORMAT_DEPTHSTENCIL) {
			viewDescr = {
				format: "depth24plus",
				aspect: "depth-only"
			};
		}
		this.view = this.createView(viewDescr);
		this.viewCache.clear();
	}
	destroy(device) {
		device.deferDestroy(this.gpuTexture);
		this.gpuTexture = null;
		this.view = null;
		this.viewCache.clear();
		this.samplers.length = 0;
	}
	propertyChanged(flag) {
		this.samplers.length = 0;
	}
	getView(device, textureView) {
		this.uploadImmediate(device, this.texture);
		if (textureView) {
			let view = this.viewCache.get(textureView.key);
			if (!view) {
				view = this.createView({
					baseMipLevel: textureView.baseMipLevel,
					mipLevelCount: textureView.mipLevelCount,
					baseArrayLayer: textureView.baseArrayLayer,
					arrayLayerCount: textureView.arrayLayerCount
				});
				this.viewCache.set(textureView.key, view);
			}
			return view;
		}
		return this.view;
	}
	createView(viewDescr) {
		const options = viewDescr ?? {};
		const textureDescr = this.desc;
		const texture = this.texture;
		const defaultViewDimension = () => {
			if (texture.cubemap) return "cube";
			if (texture.volume) return "3d";
			if (texture.array) return "2d-array";
			return "2d";
		};
		const desc = {
			format: options.format ?? textureDescr.format,
			dimension: options.dimension ?? defaultViewDimension(),
			aspect: options.aspect ?? "all",
			baseMipLevel: options.baseMipLevel ?? 0,
			mipLevelCount: options.mipLevelCount ?? textureDescr.mipLevelCount,
			baseArrayLayer: options.baseArrayLayer ?? 0,
			arrayLayerCount: options.arrayLayerCount ?? textureDescr.depthOrArrayLayers
		};
		const view = this.gpuTexture.createView(desc);
		return view;
	}
	// TODO: share a global map of samplers. Possibly even use shared samplers for bind group,
	// or maybe even have some attached in view bind group and use globally
	getSampler(device, sampleType) {
		let sampler = this.samplers[sampleType];
		if (!sampler) {
			const texture = this.texture;
			let label;
			const desc = {
				addressModeU: gpuAddressModes[texture.addressU],
				addressModeV: gpuAddressModes[texture.addressV],
				addressModeW: gpuAddressModes[texture.addressW]
			};
			if (!sampleType && texture.compareOnRead) {
				sampleType = SAMPLETYPE_DEPTH;
			}
			if (sampleType === SAMPLETYPE_DEPTH || sampleType === SAMPLETYPE_INT || sampleType === SAMPLETYPE_UINT) {
				desc.compare = "less";
				desc.magFilter = "linear";
				desc.minFilter = "linear";
				label = "Compare";
			} else if (sampleType === SAMPLETYPE_UNFILTERABLE_FLOAT) {
				desc.magFilter = "nearest";
				desc.minFilter = "nearest";
				desc.mipmapFilter = "nearest";
				label = "Unfilterable";
			} else {
				const forceNearest = !device.textureFloatFilterable && (texture.format === PIXELFORMAT_RGBA32F || texture.format === PIXELFORMAT_RGBA16F);
				if (forceNearest || this.texture.format === PIXELFORMAT_DEPTHSTENCIL || isIntegerPixelFormat(this.texture.format)) {
					desc.magFilter = "nearest";
					desc.minFilter = "nearest";
					desc.mipmapFilter = "nearest";
					label = "Nearest";
				} else {
					desc.magFilter = gpuFilterModes[texture.magFilter].level;
					desc.minFilter = gpuFilterModes[texture.minFilter].level;
					desc.mipmapFilter = gpuFilterModes[texture.minFilter].mip;
				}
			}
			const allLinear = desc.minFilter === "linear" && desc.magFilter === "linear" && desc.mipmapFilter === "linear";
			desc.maxAnisotropy = allLinear ? math.clamp(Math.round(texture._anisotropy), 1, device.maxTextureAnisotropy) : 1;
			sampler = device.wgpu.createSampler(desc);
			this.samplers[sampleType] = sampler;
		}
		return sampler;
	}
	loseContext() {
	}
	uploadImmediate(device, texture) {
		if (texture._needsUpload || texture._needsMipmapsUpload) {
			this.uploadData(device);
			texture._needsUpload = false;
			texture._needsMipmapsUpload = false;
		}
	}
	uploadData(device) {
		const texture = this.texture;
		if (this.desc && (this.desc.size.width !== texture.width || this.desc.size.height !== texture.height)) {
			device.deferDestroy(this.gpuTexture);
			this.create(device);
			texture.renderVersionDirty = device.renderVersion;
		}
		if (texture._levels) {
			let anyUploads = false;
			let anyLevelMissing = false;
			const requiredMipLevels = texture.numLevels;
			for (let mipLevel = 0; mipLevel < requiredMipLevels; mipLevel++) {
				const mipObject = texture._levels[mipLevel];
				if (mipObject) {
					if (texture.cubemap) {
						for (let face = 0; face < 6; face++) {
							const faceSource = mipObject[face];
							if (faceSource) {
								if (this.isExternalImage(faceSource)) {
									this.uploadExternalImage(device, faceSource, mipLevel, face);
									anyUploads = true;
								} else if (ArrayBuffer.isView(faceSource)) {
									this.uploadTypedArrayData(device, faceSource, mipLevel, face);
									anyUploads = true;
								} else {
								}
							} else {
								anyLevelMissing = true;
							}
						}
					} else if (texture._volume) {
					} else if (texture.array) {
						if (texture.arrayLength === mipObject.length) {
							for (let index = 0; index < texture._arrayLength; index++) {
								const arraySource = mipObject[index];
								if (this.isExternalImage(arraySource)) {
									this.uploadExternalImage(device, arraySource, mipLevel, index);
									anyUploads = true;
								} else if (ArrayBuffer.isView(arraySource)) {
									this.uploadTypedArrayData(device, arraySource, mipLevel, index);
									anyUploads = true;
								} else {
								}
							}
						} else {
							anyLevelMissing = true;
						}
					} else {
						if (device._isHTMLElementInterface(mipObject) && device.supportsHtmlTextures) {
							this.uploadElementImage(device, mipObject, mipLevel, 0);
							anyUploads = true;
						} else if (this.isExternalImage(mipObject)) {
							this.uploadExternalImage(device, mipObject, mipLevel, 0);
							anyUploads = true;
						} else if (ArrayBuffer.isView(mipObject)) {
							this.uploadTypedArrayData(device, mipObject, mipLevel, 0);
							anyUploads = true;
						} else {
						}
					}
				} else {
					anyLevelMissing = true;
				}
			}
			if (anyUploads && anyLevelMissing && texture.mipmaps && !isCompressedPixelFormat(texture.format) && !isIntegerPixelFormat(texture.format)) {
				device.mipmapRenderer.generate(this);
			}
			if (texture._gpuSize) {
				texture.adjustVramSizeTracking(device._vram, -texture._gpuSize);
			}
			texture._gpuSize = texture.gpuSize;
			texture.adjustVramSizeTracking(device._vram, texture._gpuSize);
			if (texture.releaseSourceAfterUpload) {
				texture.releaseImageSources();
			}
		}
	}
	// image types supported by copyExternalImageToTexture
	isExternalImage(image) {
		return typeof ImageBitmap !== "undefined" && image instanceof ImageBitmap || typeof HTMLVideoElement !== "undefined" && image instanceof HTMLVideoElement || typeof HTMLCanvasElement !== "undefined" && image instanceof HTMLCanvasElement || typeof OffscreenCanvas !== "undefined" && image instanceof OffscreenCanvas;
	}
	uploadExternalImage(device, image, mipLevel, index) {
		const src = {
			source: image,
			origin: [0, 0],
			flipY: false
		};
		const dst = {
			texture: this.gpuTexture,
			mipLevel,
			origin: [0, 0, index],
			aspect: "all",
			// can be: "all", "stencil-only", "depth-only"
			premultipliedAlpha: this.texture._premultiplyAlpha
		};
		const copySize = {
			width: this.desc.size.width,
			height: this.desc.size.height,
			depthOrArrayLayers: 1
			// single layer
		};
		device.submit();
		dummyUse(image instanceof HTMLCanvasElement && image.getContext("2d"));
		device.wgpu.queue.copyExternalImageToTexture(src, dst, copySize);
	}
	// upload a generic HTML element via the HTML-in-Canvas API (copyElementImageToTexture)
	uploadElementImage(device, element, mipLevel, index) {
		const dst = {
			texture: this.gpuTexture,
			mipLevel,
			origin: [0, 0, index],
			aspect: "all",
			// can be: "all", "stencil-only", "depth-only"
			premultipliedAlpha: this.texture._premultiplyAlpha
		};
		const width = TextureUtils.calcLevelDimension(this.texture.width, mipLevel);
		const height = TextureUtils.calcLevelDimension(this.texture.height, mipLevel);
		device.submit();
		const source = { source: element };
		const destination = { destination: dst, width, height };
		device.wgpu.queue.copyElementImageToTexture(source, destination);
	}
	uploadTypedArrayData(device, data, mipLevel, index) {
		const texture = this.texture;
		const wgpu = device.wgpu;
		const dest = {
			texture: this.gpuTexture,
			origin: [0, 0, index],
			mipLevel
		};
		const width = TextureUtils.calcLevelDimension(texture.width, mipLevel);
		const height = TextureUtils.calcLevelDimension(texture.height, mipLevel);
		const byteSize = TextureUtils.calcLevelGpuSize(width, height, 1, texture.format);
		const formatInfo = pixelFormatInfo.get(texture.format);
		let dataLayout;
		let size;
		if (formatInfo.size) {
			dataLayout = {
				offset: 0,
				bytesPerRow: formatInfo.size * width,
				rowsPerImage: height
			};
			size = {
				width,
				height
			};
		} else if (formatInfo.blockSize) {
			const blockDim = (size2) => {
				return Math.floor((size2 + 3) / 4);
			};
			dataLayout = {
				offset: 0,
				bytesPerRow: formatInfo.blockSize * blockDim(width),
				rowsPerImage: blockDim(height)
			};
			size = {
				width: Math.max(4, width),
				height: Math.max(4, height)
			};
		} else {
		}
		device.submit();
		wgpu.queue.writeTexture(dest, data, dataLayout, size);
	}
	read(x, y, width, height, options) {
		const mipLevel = options.mipLevel ?? 0;
		const face = options.face ?? 0;
		const data = options.data ?? null;
		const immediate = options.immediate ?? false;
		const texture = this.texture;
		const formatInfo = pixelFormatInfo.get(texture.format);
		const bytesPerRow = width * formatInfo.size;
		const paddedBytesPerRow = math.roundUp(bytesPerRow, 256);
		const size = paddedBytesPerRow * height;
		const device = texture.device;
		const stagingBuffer = device.createBufferImpl(BUFFERUSAGE_READ | BUFFERUSAGE_COPY_DST);
		stagingBuffer.allocate(device, size);
		const src = {
			texture: this.gpuTexture,
			mipLevel,
			origin: [x, y, face]
		};
		const dst = {
			buffer: stagingBuffer.buffer,
			offset: 0,
			bytesPerRow: paddedBytesPerRow
		};
		const copySize = {
			width,
			height,
			depthOrArrayLayers: 1
			// single layer
		};
		const commandEncoder = device.getCommandEncoder();
		commandEncoder.copyTextureToBuffer(src, dst, copySize);
		return device.readBuffer(stagingBuffer, size, null, immediate).then((temp) => {
			const ArrayType = getPixelFormatArrayType(texture.format);
			const targetBuffer = data?.buffer ?? new ArrayBuffer(height * bytesPerRow);
			const target = new Uint8Array(targetBuffer, data?.byteOffset ?? 0, height * bytesPerRow);
			for (let i = 0; i < height; i++) {
				const srcOffset = i * paddedBytesPerRow;
				const dstOffset = i * bytesPerRow;
				target.set(temp.subarray(srcOffset, srcOffset + bytesPerRow), dstOffset);
			}
			return data ?? new ArrayType(targetBuffer);
		});
	}
}
export {
	WebgpuTexture
};
