// playcanvas@2.19.7/build/playcanvas.prf/src/index.js
export * from "./core/constants.js";
import { extend, revision, version } from "./core/core.js";
import { guid } from "./core/guid.js";
import { path } from "./core/path.js";
import { platform } from "./core/platform.js";
import { string } from "./core/string.js";
import { EventHandler } from "./core/event-handler.js";
import { EventHandle } from "./core/event-handle.js";
import { IndexedList } from "./core/indexed-list.js";
import { WasmModule } from "./core/wasm-module.js";
import { ReadStream } from "./core/read-stream.js";
import { SortedLoopArray } from "./core/sorted-loop-array.js";
import { Tags } from "./core/tags.js";
import { now } from "./core/time.js";
import { URI, createURI } from "./core/uri.js";
import { Tracing } from "./core/tracing.js";
export * from "./core/math/constants.js";
import { math } from "./core/math/math.js";
import { Color } from "./core/math/color.js";
import { Curve } from "./core/math/curve.js";
import { CurveSet } from "./core/math/curve-set.js";
import { FloatPacking } from "./core/math/float-packing.js";
import { Kernel } from "./core/math/kernel.js";
import { Mat3 } from "./core/math/mat3.js";
import { Mat4 } from "./core/math/mat4.js";
import { Quat } from "./core/math/quat.js";
import { Vec2 } from "./core/math/vec2.js";
import { Vec3 } from "./core/math/vec3.js";
import { Vec4 } from "./core/math/vec4.js";
import { BoundingBox } from "./core/shape/bounding-box.js";
import { BoundingSphere } from "./core/shape/bounding-sphere.js";
import { Frustum } from "./core/shape/frustum.js";
import { OrientedBox } from "./core/shape/oriented-box.js";
import { Plane } from "./core/shape/plane.js";
import { Tri } from "./core/shape/tri.js";
import { Ray } from "./core/shape/ray.js";
export * from "./platform/graphics/constants.js";
import { createGraphicsDevice } from "./platform/graphics/graphics-device-create.js";
import { BindGroupFormat, BindUniformBufferFormat, BindTextureFormat, BindStorageTextureFormat, BindStorageBufferFormat } from "./platform/graphics/bind-group-format.js";
import { BlendState } from "./platform/graphics/blend-state.js";
import { Compute } from "./platform/graphics/compute.js";
import { DepthState } from "./platform/graphics/depth-state.js";
import { DrawCommands } from "./platform/graphics/draw-commands.js";
import { GraphicsDevice } from "./platform/graphics/graphics-device.js";
import { IndexBuffer } from "./platform/graphics/index-buffer.js";
import { RenderTarget } from "./platform/graphics/render-target.js";
import { RenderPass } from "./platform/graphics/render-pass.js";
import { ScopeId } from "./platform/graphics/scope-id.js";
import { ScopeSpace } from "./platform/graphics/scope-space.js";
import { Shader } from "./platform/graphics/shader.js";
import { StorageBuffer } from "./platform/graphics/storage-buffer.js";
import { Texture } from "./platform/graphics/texture.js";
import { TextureUtils } from "./platform/graphics/texture-utils.js";
import { TextureView } from "./platform/graphics/texture-view.js";
import { TransformFeedback } from "./platform/graphics/transform-feedback.js";
import { UniformBufferFormat, UniformFormat } from "./platform/graphics/uniform-buffer-format.js";
import { VertexBuffer } from "./platform/graphics/vertex-buffer.js";
import { VertexFormat } from "./platform/graphics/vertex-format.js";
import { VertexIterator } from "./platform/graphics/vertex-iterator.js";
import { WebglGraphicsDevice } from "./platform/graphics/webgl/webgl-graphics-device.js";
import { WebgpuGraphicsDevice } from "./platform/graphics/webgpu/webgpu-graphics-device.js";
import { NullGraphicsDevice } from "./platform/graphics/null/null-graphics-device.js";
export * from "./platform/input/constants.js";
import { Controller } from "./platform/input/controller.js";
import { GamePads } from "./platform/input/game-pads.js";
import { Keyboard } from "./platform/input/keyboard.js";
import { KeyboardEvent } from "./platform/input/keyboard-event.js";
import { Mouse } from "./platform/input/mouse.js";
import { MouseEvent } from "./platform/input/mouse-event.js";
import { TouchDevice } from "./platform/input/touch-device.js";
import { getTouchTargetCoords, Touch, TouchEvent } from "./platform/input/touch-event.js";
import { http, Http } from "./platform/net/http.js";
export * from "./platform/sound/constants.js";
import { SoundManager } from "./platform/sound/manager.js";
import { Sound } from "./platform/sound/sound.js";
import { SoundInstance } from "./platform/sound/instance.js";
import { SoundInstance3d } from "./platform/sound/instance3d.js";
export * from "./scene/constants.js";
import { drawQuadWithShader } from "./scene/graphics/quad-render-utils.js";
import { Batch } from "./scene/batching/batch.js";
import { BatchGroup } from "./scene/batching/batch-group.js";
import { SkinBatchInstance } from "./scene/batching/skin-batch-instance.js";
import { BatchManager } from "./scene/batching/batch-manager.js";
import { Camera } from "./scene/camera.js";
import { WorldClusters } from "./scene/lighting/world-clusters.js";
import { ForwardRenderer } from "./scene/renderer/forward-renderer.js";
import { GraphNode } from "./scene/graph-node.js";
import { Layer } from "./scene/layer.js";
import { LayerComposition } from "./scene/composition/layer-composition.js";
import { Light } from "./scene/light.js";
import { LightingParams } from "./scene/lighting/lighting-params.js";
import { Mesh } from "./scene/mesh.js";
import { MeshInstance } from "./scene/mesh-instance.js";
import { Model } from "./scene/model.js";
import { Morph } from "./scene/morph.js";
import { MorphInstance } from "./scene/morph-instance.js";
import { MorphTarget } from "./scene/morph-target.js";
import { ParticleEmitter } from "./scene/particle-system/particle-emitter.js";
import { QuadRender } from "./scene/graphics/quad-render.js";
import { Scene } from "./scene/scene.js";
import { ShaderPass } from "./scene/shader-pass.js";
import { Skin } from "./scene/skin.js";
import { SkinInstance } from "./scene/skin-instance.js";
import { Sprite } from "./scene/sprite.js";
import { StencilParameters } from "./platform/graphics/stencil-parameters.js";
import { TextureAtlas } from "./scene/texture-atlas.js";
import { Animation, AnimationKey, AnimationNode } from "./scene/animation/animation.js";
import { Skeleton } from "./scene/animation/skeleton.js";
import { ComputeRadixSort } from "./scene/graphics/radix-sort/compute-radix-sort.js";
import { EnvLighting } from "./scene/graphics/env-lighting.js";
import { PostEffect } from "./scene/graphics/post-effect.js";
import { FramePass } from "./platform/graphics/frame-pass.js";
import { FramePassColorGrab } from "./scene/graphics/frame-pass-color-grab.js";
import { FramePassRadixSort } from "./scene/graphics/frame-pass-radix-sort.js";
import { RenderPassShaderQuad } from "./scene/graphics/render-pass-shader-quad.js";
import { reprojectTexture } from "./scene/graphics/reproject-texture.js";
import { LitMaterial } from "./scene/materials/lit-material.js";
import { Material } from "./scene/materials/material.js";
import { ShaderMaterial } from "./scene/materials/shader-material.js";
import { StandardMaterial } from "./scene/materials/standard-material.js";
import { StandardMaterialOptions } from "./scene/materials/standard-material-options.js";
import { calculateNormals, calculateTangents } from "./scene/geometry/geometry-utils.js";
import { CapsuleGeometry } from "./scene/geometry/capsule-geometry.js";
import { ConeGeometry } from "./scene/geometry/cone-geometry.js";
import { CylinderGeometry } from "./scene/geometry/cylinder-geometry.js";
import { DomeGeometry } from "./scene/geometry/dome-geometry.js";
import { Geometry } from "./scene/geometry/geometry.js";
import { BoxGeometry } from "./scene/geometry/box-geometry.js";
import { PlaneGeometry } from "./scene/geometry/plane-geometry.js";
import { SphereGeometry } from "./scene/geometry/sphere-geometry.js";
import { TorusGeometry } from "./scene/geometry/torus-geometry.js";
import { FogParams } from "./scene/fog-params.js";
import { RenderPassForward } from "./scene/renderer/render-pass-forward.js";
import { ShaderUtils, createShader, createShaderFromCode } from "./scene/shader-lib/shader-utils.js";
import { LitShaderOptions } from "./scene/shader-lib/programs/lit-shader-options.js";
import { ProgramLibrary } from "./scene/shader-lib/program-library.js";
import { ChunkUtils } from "./scene/shader-lib/chunk-utils.js";
import { ShaderChunks } from "./scene/shader-lib/shader-chunks.js";
import { Sky } from "./scene/skybox/sky.js";
import { GSplatContainer } from "./scene/gsplat/gsplat-container.js";
import { GSplatData } from "./scene/gsplat/gsplat-data.js";
import { GSplatFormat } from "./scene/gsplat/gsplat-format.js";
import { GSplatInstance } from "./scene/gsplat/gsplat-instance.js";
import { GSplatProcessor } from "./framework/gsplat/gsplat-processor.js";
import { GSplatResource } from "./scene/gsplat/gsplat-resource.js";
import { GSplatResourceBase } from "./scene/gsplat/gsplat-resource-base.js";
import { GSplatSogData } from "./scene/gsplat/gsplat-sog-data.js";
import { GSplatSogResource } from "./scene/gsplat/gsplat-sog-resource.js";
export * from "./framework/constants.js";
import { script } from "./framework/script.js";
import { AppBase, app } from "./framework/app-base.js";
import { AppOptions } from "./framework/app-options.js";
import { Application } from "./framework/application.js";
import { AnimationComponent } from "./framework/components/animation/component.js";
import { AnimationComponentSystem } from "./framework/components/animation/system.js";
import { AnimComponent } from "./framework/components/anim/component.js";
import { AnimComponentLayer } from "./framework/components/anim/component-layer.js";
import { AnimComponentSystem } from "./framework/components/anim/system.js";
import { AudioListenerComponent } from "./framework/components/audio-listener/component.js";
import { AudioListenerComponentSystem } from "./framework/components/audio-listener/system.js";
export * from "./framework/components/button/constants.js";
import { ButtonComponent } from "./framework/components/button/component.js";
import { ButtonComponentSystem } from "./framework/components/button/system.js";
import { CameraComponent } from "./framework/components/camera/component.js";
import { CameraComponentSystem } from "./framework/components/camera/system.js";
import { CollisionComponent } from "./framework/components/collision/component.js";
import { CollisionComponentSystem } from "./framework/components/collision/system.js";
import { Component } from "./framework/components/component.js";
import { ComponentSystem } from "./framework/components/system.js";
import { ComponentSystemRegistry } from "./framework/components/registry.js";
export * from "./framework/components/element/constants.js";
import { ElementComponent } from "./framework/components/element/component.js";
import { ElementComponentSystem } from "./framework/components/element/system.js";
import { ElementDragHelper } from "./framework/components/element/element-drag-helper.js";
import { Entity } from "./framework/entity.js";
import { GSplatComponent } from "./framework/components/gsplat/component.js";
import { GSplatComponentSystem } from "./framework/components/gsplat/system.js";
import { ImageElement } from "./framework/components/element/image-element.js";
export * from "./framework/components/joint/constants.js";
import { JointComponent } from "./framework/components/joint/component.js";
import { JointComponentSystem } from "./framework/components/joint/system.js";
import { LayoutCalculator } from "./framework/components/layout-group/layout-calculator.js";
import { LayoutChildComponent } from "./framework/components/layout-child/component.js";
import { LayoutChildComponentSystem } from "./framework/components/layout-child/system.js";
export * from "./framework/components/layout-group/constants.js";
import { LayoutGroupComponent } from "./framework/components/layout-group/component.js";
import { LayoutGroupComponentSystem } from "./framework/components/layout-group/system.js";
import { LightComponent } from "./framework/components/light/component.js";
import { LightComponentSystem } from "./framework/components/light/system.js";
import { Lightmapper } from "./framework/lightmapper/lightmapper.js";
import { ModelComponent } from "./framework/components/model/component.js";
import { ModelComponentSystem } from "./framework/components/model/system.js";
import { ParticleSystemComponent } from "./framework/components/particle-system/component.js";
import { ParticleSystemComponentSystem } from "./framework/components/particle-system/system.js";
import { PostEffectQueue } from "./framework/components/camera/post-effect-queue.js";
import { RenderComponent } from "./framework/components/render/component.js";
import { RenderComponentSystem } from "./framework/components/render/system.js";
export * from "./framework/components/rigid-body/constants.js";
import { RigidBodyComponent } from "./framework/components/rigid-body/component.js";
import { RigidBodyComponentSystem, ContactPoint, ContactResult, RaycastResult, SingleContactResult } from "./framework/components/rigid-body/system.js";
import { SceneRegistry } from "./framework/scene-registry.js";
import { SceneRegistryItem } from "./framework/scene-registry-item.js";
export * from "./framework/components/screen/constants.js";
import { ScreenComponent } from "./framework/components/screen/component.js";
import { ScreenComponentSystem } from "./framework/components/screen/system.js";
import { ScriptComponent } from "./framework/components/script/component.js";
import { ScriptComponentSystem } from "./framework/components/script/system.js";
import { ScrollbarComponent } from "./framework/components/scrollbar/component.js";
import { ScrollbarComponentSystem } from "./framework/components/scrollbar/system.js";
export * from "./framework/components/scroll-view/constants.js";
import { ScrollViewComponent } from "./framework/components/scroll-view/component.js";
import { ScrollViewComponentSystem } from "./framework/components/scroll-view/system.js";
import { SoundSlot } from "./framework/components/sound/slot.js";
import { SoundComponent } from "./framework/components/sound/component.js";
import { SoundComponentSystem } from "./framework/components/sound/system.js";
export * from "./framework/components/sprite/constants.js";
import { SpriteAnimationClip } from "./framework/components/sprite/sprite-animation-clip.js";
import { SpriteComponent } from "./framework/components/sprite/component.js";
import { SpriteComponentSystem } from "./framework/components/sprite/system.js";
import { Template } from "./framework/template.js";
import { TextElement } from "./framework/components/element/text-element.js";
import { ZoneComponent } from "./framework/components/zone/component.js";
import { ZoneComponentSystem } from "./framework/components/zone/system.js";
export * from "./framework/anim/constants.js";
import { AnimBinder } from "./framework/anim/binder/anim-binder.js";
import { AnimClip } from "./framework/anim/evaluator/anim-clip.js";
import { AnimCurve } from "./framework/anim/evaluator/anim-curve.js";
import { AnimData } from "./framework/anim/evaluator/anim-data.js";
import { AnimEvaluator } from "./framework/anim/evaluator/anim-evaluator.js";
import { AnimSnapshot } from "./framework/anim/evaluator/anim-snapshot.js";
import { AnimTarget } from "./framework/anim/evaluator/anim-target.js";
import { AnimEvents } from "./framework/anim/evaluator/anim-events.js";
import { AnimTrack } from "./framework/anim/evaluator/anim-track.js";
import { DefaultAnimBinder } from "./framework/anim/binder/default-anim-binder.js";
export * from "./framework/anim/controller/constants.js";
import { AnimController } from "./framework/anim/controller/anim-controller.js";
import { AnimStateGraph } from "./framework/anim/state-graph/anim-state-graph.js";
export * from "./framework/asset/constants.js";
import { Asset } from "./framework/asset/asset.js";
import { AssetListLoader } from "./framework/asset/asset-list-loader.js";
import { AssetReference } from "./framework/asset/asset-reference.js";
import { AssetRegistry } from "./framework/asset/asset-registry.js";
import { LocalizedAsset } from "./framework/asset/asset-localized.js";
export * from "./framework/font/constants.js";
import { Font } from "./framework/font/font.js";
import { CanvasFont } from "./framework/font/canvas-font.js";
import { Bundle } from "./framework/bundle/bundle.js";
import { BundleRegistry } from "./framework/bundle/bundle-registry.js";
import { Picker } from "./framework/graphics/picker.js";
import { RenderPassPicker } from "./framework/graphics/render-pass-picker.js";
import { basisInitialize } from "./framework/handlers/basis.js";
import { dracoInitialize, dracoDecode } from "./framework/parsers/draco-decoder.js";
import { AnimClipHandler } from "./framework/handlers/anim-clip.js";
import { AnimStateGraphHandler } from "./framework/handlers/anim-state-graph.js";
import { AnimationHandler } from "./framework/handlers/animation.js";
import { AudioHandler } from "./framework/handlers/audio.js";
import { BinaryHandler } from "./framework/handlers/binary.js";
import { BundleHandler } from "./framework/handlers/bundle.js";
import { ContainerHandler, ContainerResource } from "./framework/handlers/container.js";
import { CssHandler } from "./framework/handlers/css.js";
import { CubemapHandler } from "./framework/handlers/cubemap.js";
import { FolderHandler } from "./framework/handlers/folder.js";
import { FontHandler } from "./framework/handlers/font.js";
import { HierarchyHandler } from "./framework/handlers/hierarchy.js";
import { HtmlHandler } from "./framework/handlers/html.js";
import { JsonHandler } from "./framework/handlers/json.js";
import { MaterialHandler } from "./framework/handlers/material.js";
import { ModelHandler } from "./framework/handlers/model.js";
import { GSplatHandler } from "./framework/handlers/gsplat.js";
import { RenderHandler } from "./framework/handlers/render.js";
import { ResourceHandler } from "./framework/handlers/handler.js";
import { ResourceLoader } from "./framework/handlers/loader.js";
import { ScriptHandler } from "./framework/handlers/script.js";
import { SceneHandler } from "./framework/handlers/scene.js";
import { SceneSettingsHandler } from "./framework/handlers/scene-settings.js";
import { ShaderHandler } from "./framework/handlers/shader.js";
import { SpriteHandler } from "./framework/handlers/sprite.js";
import { TemplateHandler } from "./framework/handlers/template.js";
import { TextHandler } from "./framework/handlers/text.js";
import { TextureHandler } from "./framework/handlers/texture.js";
import { TextureAtlasHandler } from "./framework/handlers/texture-atlas.js";
import { ElementInput, ElementInputEvent, ElementMouseEvent, ElementSelectEvent, ElementTouchEvent } from "./framework/input/element-input.js";
import { JsonStandardMaterialParser } from "./framework/parsers/material/json-standard-material.js";
import { createScript, registerScript, getReservedScriptNames } from "./framework/script/script-create.js";
import { ScriptAttributes } from "./framework/script/script-attributes.js";
import { ScriptRegistry } from "./framework/script/script-registry.js";
import { ScriptType } from "./framework/script/script-type.js";
import { Script } from "./framework/script/script.js";
import { I18n } from "./framework/i18n/i18n.js";
export * from "./framework/xr/constants.js";
import { XrAnchor } from "./framework/xr/xr-anchor.js";
import { XrAnchors } from "./framework/xr/xr-anchors.js";
import { XrDomOverlay } from "./framework/xr/xr-dom-overlay.js";
import { XrFinger } from "./framework/xr/xr-finger.js";
import { XrHand } from "./framework/xr/xr-hand.js";
import { XrHitTest } from "./framework/xr/xr-hit-test.js";
import { XrHitTestSource } from "./framework/xr/xr-hit-test-source.js";
import { XrImageTracking } from "./framework/xr/xr-image-tracking.js";
import { XrInput } from "./framework/xr/xr-input.js";
import { XrInputSource } from "./framework/xr/xr-input-source.js";
import { XrJoint } from "./framework/xr/xr-joint.js";
import { XrLightEstimation } from "./framework/xr/xr-light-estimation.js";
import { XrManager } from "./framework/xr/xr-manager.js";
import { XrMeshDetection } from "./framework/xr/xr-mesh-detection.js";
import { XrPlane } from "./framework/xr/xr-plane.js";
import { XrPlaneDetection } from "./framework/xr/xr-plane-detection.js";
import { XrTrackedImage } from "./framework/xr/xr-tracked-image.js";
import { XrView } from "./framework/xr/xr-view.js";
import { XrViews } from "./framework/xr/xr-views.js";
export * from "./deprecated/deprecated.js";
export * from "./extras/index.js";
export {
	AnimBinder,
	AnimClip,
	AnimClipHandler,
	AnimComponent,
	AnimComponentLayer,
	AnimComponentSystem,
	AnimController,
	AnimCurve,
	AnimData,
	AnimEvaluator,
	AnimEvents,
	AnimSnapshot,
	AnimStateGraph,
	AnimStateGraphHandler,
	AnimTarget,
	AnimTrack,
	Animation,
	AnimationComponent,
	AnimationComponentSystem,
	AnimationHandler,
	AnimationKey,
	AnimationNode,
	AppBase,
	AppOptions,
	Application,
	Asset,
	AssetListLoader,
	AssetReference,
	AssetRegistry,
	AudioHandler,
	AudioListenerComponent,
	AudioListenerComponentSystem,
	Batch,
	BatchGroup,
	BatchManager,
	BinaryHandler,
	BindGroupFormat,
	BindStorageBufferFormat,
	BindStorageTextureFormat,
	BindTextureFormat,
	BindUniformBufferFormat,
	BlendState,
	BoundingBox,
	BoundingSphere,
	BoxGeometry,
	Bundle,
	BundleHandler,
	BundleRegistry,
	ButtonComponent,
	ButtonComponentSystem,
	Camera,
	CameraComponent,
	CameraComponentSystem,
	CanvasFont,
	CapsuleGeometry,
	ChunkUtils,
	CollisionComponent,
	CollisionComponentSystem,
	Color,
	Component,
	ComponentSystem,
	ComponentSystemRegistry,
	Compute,
	ComputeRadixSort,
	ConeGeometry,
	ContactPoint,
	ContactResult,
	ContainerHandler,
	ContainerResource,
	Controller,
	CssHandler,
	CubemapHandler,
	Curve,
	CurveSet,
	CylinderGeometry,
	DefaultAnimBinder,
	DepthState,
	DomeGeometry,
	DrawCommands,
	ElementComponent,
	ElementComponentSystem,
	ElementDragHelper,
	ElementInput,
	ElementInputEvent,
	ElementMouseEvent,
	ElementSelectEvent,
	ElementTouchEvent,
	Entity,
	EnvLighting,
	EventHandle,
	EventHandler,
	FloatPacking,
	FogParams,
	FolderHandler,
	Font,
	FontHandler,
	ForwardRenderer,
	FramePass,
	FramePassColorGrab,
	FramePassRadixSort,
	Frustum,
	GSplatComponent,
	GSplatComponentSystem,
	GSplatContainer,
	GSplatData,
	GSplatFormat,
	GSplatHandler,
	GSplatInstance,
	GSplatProcessor,
	GSplatResource,
	GSplatResourceBase,
	GSplatSogData,
	GSplatSogResource,
	GamePads,
	Geometry,
	GraphNode,
	GraphicsDevice,
	HierarchyHandler,
	HtmlHandler,
	Http,
	I18n,
	ImageElement,
	IndexBuffer,
	IndexedList,
	JointComponent,
	JointComponentSystem,
	JsonHandler,
	JsonStandardMaterialParser,
	Kernel,
	Keyboard,
	KeyboardEvent,
	Layer,
	LayerComposition,
	LayoutCalculator,
	LayoutChildComponent,
	LayoutChildComponentSystem,
	LayoutGroupComponent,
	LayoutGroupComponentSystem,
	Light,
	LightComponent,
	LightComponentSystem,
	LightingParams,
	Lightmapper,
	LitMaterial,
	LitShaderOptions,
	LocalizedAsset,
	Mat3,
	Mat4,
	Material,
	MaterialHandler,
	Mesh,
	MeshInstance,
	Model,
	ModelComponent,
	ModelComponentSystem,
	ModelHandler,
	Morph,
	MorphInstance,
	MorphTarget,
	Mouse,
	MouseEvent,
	NullGraphicsDevice,
	OrientedBox,
	ParticleEmitter,
	ParticleSystemComponent,
	ParticleSystemComponentSystem,
	Picker,
	Plane,
	PlaneGeometry,
	PostEffect,
	PostEffectQueue,
	ProgramLibrary,
	QuadRender,
	Quat,
	Ray,
	RaycastResult,
	ReadStream,
	RenderComponent,
	RenderComponentSystem,
	RenderHandler,
	RenderPass,
	RenderPassForward,
	RenderPassPicker,
	RenderPassShaderQuad,
	RenderTarget,
	ResourceHandler,
	ResourceLoader,
	RigidBodyComponent,
	RigidBodyComponentSystem,
	Scene,
	SceneHandler,
	SceneRegistry,
	SceneRegistryItem,
	SceneSettingsHandler,
	ScopeId,
	ScopeSpace,
	ScreenComponent,
	ScreenComponentSystem,
	Script,
	ScriptAttributes,
	ScriptComponent,
	ScriptComponentSystem,
	ScriptHandler,
	ScriptRegistry,
	ScriptType,
	ScrollViewComponent,
	ScrollViewComponentSystem,
	ScrollbarComponent,
	ScrollbarComponentSystem,
	Shader,
	ShaderChunks,
	ShaderHandler,
	ShaderMaterial,
	ShaderPass,
	ShaderUtils,
	SingleContactResult,
	Skeleton,
	Skin,
	SkinBatchInstance,
	SkinInstance,
	Sky,
	SortedLoopArray,
	Sound,
	SoundComponent,
	SoundComponentSystem,
	SoundInstance,
	SoundInstance3d,
	SoundManager,
	SoundSlot,
	SphereGeometry,
	Sprite,
	SpriteAnimationClip,
	SpriteComponent,
	SpriteComponentSystem,
	SpriteHandler,
	StandardMaterial,
	StandardMaterialOptions,
	StencilParameters,
	StorageBuffer,
	Tags,
	Template,
	TemplateHandler,
	TextElement,
	TextHandler,
	Texture,
	TextureAtlas,
	TextureAtlasHandler,
	TextureHandler,
	TextureUtils,
	TextureView,
	TorusGeometry,
	Touch,
	TouchDevice,
	TouchEvent,
	Tracing,
	TransformFeedback,
	Tri,
	URI,
	UniformBufferFormat,
	UniformFormat,
	Vec2,
	Vec3,
	Vec4,
	VertexBuffer,
	VertexFormat,
	VertexIterator,
	WasmModule,
	WebglGraphicsDevice,
	WebgpuGraphicsDevice,
	WorldClusters,
	XrAnchor,
	XrAnchors,
	XrDomOverlay,
	XrFinger,
	XrHand,
	XrHitTest,
	XrHitTestSource,
	XrImageTracking,
	XrInput,
	XrInputSource,
	XrJoint,
	XrLightEstimation,
	XrManager,
	XrMeshDetection,
	XrPlane,
	XrPlaneDetection,
	XrTrackedImage,
	XrView,
	XrViews,
	ZoneComponent,
	ZoneComponentSystem,
	app,
	basisInitialize,
	calculateNormals,
	calculateTangents,
	createGraphicsDevice,
	createScript,
	createShader,
	createShaderFromCode,
	createURI,
	dracoDecode,
	dracoInitialize,
	drawQuadWithShader,
	extend,
	getReservedScriptNames,
	getTouchTargetCoords,
	guid,
	http,
	math,
	now,
	path,
	platform,
	registerScript,
	reprojectTexture,
	revision,
	script,
	string,
	version
};
